### PR TITLE
Update vhost-user-backend to 0.11 series

### PR DESCRIPTION
**Pull request by Ablu**
- features were renamed from slave -> backend
- generics got simplified

Draft until the version is actually released

### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [x] All commits in this PR are signed (with `git commit -s`), and the commit
  message has max 60 characters for the summary and max 75 characters for each
  description line.
- [x] All added/changed functionality has a corresponding unit/integration
  test.
- [x] All added/changed public-facing functionality has entries in the "Upcoming 
  Release" section of CHANGELOG.md (if no such section exists, please create one).
- [x] Any newly added `unsafe` code is properly documented.

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -4,18 +4,18 @@ version = 3
 
 [[package]]
 name = "aho-corasick"
-version = "1.0.5"
+version = "1.1.1"
```
Comment (stefano-garzarella) [inline Cargo.lock:7]: I would suggest updating just vhost and vhost-user-backend: `cargo update -p vhost -p vhost-user-backend`, but we can fix later ;-)

Thanks for this PR!

Comment (Ablu) [reply Cargo.lock:7]: > I would suggest updating just vhost and vhost-user-backend: `cargo update -p vhost -p vhost-user-backend`, but we can fix later ;-)

I would add `--aggressive` to also update whatever those crates depend on, but yeah. I will clean that once the release is there.

**Comment by Ablu** [comment]
Hm... We need another virtio-vsock update as well...

**Comment by Ablu** [comment]
Hm... vsock_conn.rs tests could probably benefit from a WriteVolatile impl on Cursor<Vec>. But I am not going through another round of major updates for that. Will post the patches to add it in vm-memory but then do an additional copy for the test code here. Once the vm-memory update then goes through we can drop it.

**Comment by stefano-garzarella** [comment]
> Hm... vsock_conn.rs tests could probably benefit from a WriteVolatile impl on Cursor. But I am not going through another round of major updates for that. Will post the patches to add it in vm-memory but then do an additional copy for the test code here. Once the vm-memory update then goes through we can drop it.

Yep, I agree!

**Comment by Ablu** [comment]
This is ready now! The vsock tests gave me some headache and went back and forth a couple of times until I came up with this. Since the VecDequeue impls cannot be done very nicely, I am not sure how much value there is in trying to upstream them to vm-memory. The tricks that we did to make it fast for `Vec` do not work here...

**Comment by Ablu** [comment]
> This is ready now!

Except that it needs a rebase now...

**Comment by Ablu** [comment]
Rebased.

**Comment by Ablu** [comment]
Ah. Forgot about staging...

File: vhost-device-vsock/src/vhu_vsock_thread.rs [hunk 2]
```diff
@@ -242,7 +242,7 @@ impl VhostUserVsockThread {
     /// Register our listeners in the VringEpollHandler
     pub fn register_listeners(
         &mut self,
-        epoll_handler: Arc<VringEpollHandler<ArcVhostBknd, VringRwLock, ()>>,
+        epoll_handler: Arc<VringEpollHandler<ArcVhostBknd>>,
```
Comment (stefano-garzarella) [inline vhost-device-vsock/src/vhu_vsock_thread.rs:245]: `cargo fmt` CI failure here

File: staging/vhost-device-sound/Cargo.toml [hunk 3]
```diff
@@ -26,12 +26,12 @@ thiserror = "1.0"
 vhost = { version = "0.8", features = ["vhost-user-slave"] }
 vhost-user-backend = "0.10"
 virtio-bindings = "0.2.1"
-virtio-queue = "0.9"
-vm-memory = "0.12"
+virtio-queue = "0.10"
+vm-memory = "0.13"
```
Comment (stefano-garzarella) [inline staging/vhost-device-sound/Cargo.toml:30]: Should we use `0.13.1`?

Comment (Ablu) [reply staging/vhost-device-sound/Cargo.toml:30]: Yes... That one brings the missing traits. Will update

**Review by stefano-garzarella — COMMENTED** [review]
LGTM! Thanks for taking care of vsock tests!

I left some minor comments.

**Comment by stefano-garzarella** [comment]
@Ablu staging crates are still failing: https://buildkite.com/rust-vmm/vhost-device-ci/builds/1960#018bc957-95ab-48d2-bea7-00fea0939a0a
Is it related?

**Comment by Ablu** [comment]
> @Ablu staging crates are still failing: https://buildkite.com/rust-vmm/vhost-device-ci/builds/1960#018bc957-95ab-48d2-bea7-00fea0939a0a
> Is it related?

Yes, still fixing it. forgot to bump backend, but something else is missing too. Will sort it out.

**Comment by Ablu** [comment]
Hm. Is this crash known for staging?

```
Caused by:
  process didn't exit successfully: `/home/ablu/projects/rust-vmm/vhost-device/staging/target/debug/deps/vhost_device_sound-45dd1b9dd31276f6` (signal: 11, SIGSEGV: invalid memory reference)
```

**Comment by Ablu** [comment]
Apart from the crash I also finished the update in staging.

**Comment by stefano-garzarella** [comment]
> Hm. Is this crash known for staging?
> 
> ```
> Caused by:
>   process didn't exit successfully: `/home/ablu/projects/rust-vmm/vhost-device/staging/target/debug/deps/vhost_device_sound-45dd1b9dd31276f6` (signal: 11, SIGSEGV: invalid memory reference)
> ```

@dorindabassey @MatiasVara @epilys any idea of this?

**Comment by MatiasVara** [comment]
> > Hm. Is this crash known for staging?
> > ```
> > Caused by:
> >   process didn't exit successfully: `/home/ablu/projects/rust-vmm/vhost-device/staging/target/debug/deps/vhost_device_sound-45dd1b9dd31276f6` (signal: 11, SIGSEGV: invalid memory reference)
> > ```
> 
> @dorindabassey @MatiasVara @epilys any idea of this?

Those logs are about `virtio-device-video` if I am not wrong.

**Comment by stefano-garzarella** [comment]
> > > Hm. Is this crash known for staging?
> > > ```
> > > Caused by:
> > >   process didn't exit successfully: `/home/ablu/projects/rust-vmm/vhost-device/staging/target/debug/deps/vhost_device_sound-45dd1b9dd31276f6` (signal: 11, SIGSEGV: invalid memory reference)
> > > ```
> > 
> > 
> > @dorindabassey @MatiasVara @epilys any idea of this?
> 
> Those logs are about `virtio-device-video` if I am not wrong.

The executable looks like the sound one: `vhost_device_sound-45dd1b9dd31276f6`

File: staging/vhost-device-sound/src/device.rs [hunk 4]
```diff
@@ -1111,16 +1114,16 @@ mod tests {
...
-        assert!(!ret.unwrap());
+        assert!(ret.is_ok());
 
         let ret = backend.handle_event(TX_QUEUE_IDX, EventSet::IN, &vrings, 0);
-        assert!(!ret.unwrap());
+        assert!(ret.is_ok());
 
         let ret = backend.handle_event(RX_QUEUE_IDX, EventSet::IN, &vrings, 0);
-        assert!(!ret.unwrap());
+        assert!(ret.is_ok());
```
Comment (stefano-garzarella) [inline staging/vhost-device-sound/src/device.rs:1126]: Clippy is complaining about these changes, see https://buildkite.com/rust-vmm/vhost-device-ci/builds/1962#018bc964-fa22-4499-a69c-b8bdd330cf27

It is suggesting to use just `ret.unwrap()`.

Comment (Ablu) [reply staging/vhost-device-sound/src/device.rs]: fixed... need to update my test.sh script with staging/...

**Comment by Ablu** [comment]
I can no longer reproduce the crash during tests... So I guess that can be handled separately from this. I will try to reproduce it with address sanitizers and open an issue if I can reproduce it again.

**Review by epilys — APPROVED** [review]

**Comment by epilys** [comment]
Is it okay with everyone to delay merging till we test some of the devices by launching them with VMs? Just to make sure everything is ok to go.

**Comment by Ablu** [comment]
> Is it okay with everyone to delay merging till we test some of the devices by launching them with VMs? Just to make sure everything is ok to go.

Sure! Though we probably have a testing gap somewhere if stuff is failing there but not in our tests.

**Comment by stefano-garzarella** [comment]
I reproduced the `vhost-device-sound` issue also on `main`, so I don't think it is related to this PR. For me, we can go head merging this.

**Review by stefano-garzarella — APPROVED** [review]

**Comment by Ablu** [comment]
Thanks for reviews!

@epilys: Got any specific tests in mind? Feel free to hit merge when you are happy.

I am having better integration tests somewhere on my TODO list. Though I am not yet sure about the final form... We will probably develop integration tests for [TRS](https://linaro.atlassian.net/wiki/spaces/TRS/overview), but those would only be able to run post-merge.

Maybe it might make sense to collect ideas on a ticket? Overall, I think this is not more or less dangerous compared to any other PR. While this includes some major bumps, the actual changes are fairly "simple".

**Pull request merged**